### PR TITLE
Annotate Live Plan with additional node info

### PR DESCRIPTION
**Pull request by raghavsethi**
Variant of #11927 which allows the output to be even more structured.

This change paves the way for PlanPrinter to generate more structured
output than a text blob. This structured output can then be used to
enrich a raw JSON plan.

Test plan:
- Run AbstractTestQueries with patch that dumps plan for every query, compare two dozen or so query plans with master.
- Run AbstractTestQueries with patch that does not catch exceptions thrown when constructing text plans, and verify that tests all pass.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 1]
```diff
@@ -119,28 +113,25 @@
...
-import static com.google.common.collect.Iterables.getOnlyElement;
-import static io.airlift.units.DataSize.succinctBytes;
-import static java.lang.Double.isFinite;
-import static java.lang.Double.isNaN;
 import static java.lang.String.format;
-import static java.util.Arrays.stream;
 import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.joining;
-import static java.util.stream.Collectors.toList;
+import static org.weakref.jmx.internal.guava.base.Preconditions.checkState;
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: can we have some kind of a check for this?

[the remark this answers is not in the document]
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: This failed checkstyle - I pushed a bad commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PrintOutput.java [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+import static com.facebook.presto.sql.planner.planPrinter.PrintUtils.formatOutputs;
+import static com.facebook.presto.sql.planner.planPrinter.PrintUtils.formatPositions;
+import static com.facebook.presto.sql.planner.planPrinter.PrintUtils.indentString;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+public class PrintOutput
```
[the remark this answers is not in the document]
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PrintOutput.java]: Next step would be to create a class the stores all the details for the node, and then pass that down to the UIs.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PrintOutput.java [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        PlanNodeStats nodeStats = stats.get().get(rootNode.getId());
+        if (nodeStats == null) {
+            output.append("Cost: ?, Output: ? rows (?B)\n");
+            return;
+        }
+
+        double scheduledTimeFraction = 100.0d * nodeStats.getPlanNodeScheduledTime().toMillis() / totalScheduledMillis;
+        double cpuTimeFraction = 100.0d * nodeStats.getPlanNodeCpuTime().toMillis() / totalCpuMillis;
+
+        output.append(format("CPU: %s (%s%%), Scheduled: %s (%s%%)",
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PrintOutput.java]: On L174 above and maybe in other places `cpu` is lowercase, here we use it in caps. It would be nice to be consistent.

**Comment by raghavsethi** [comment]
I'm redoing parts of this after a conversation with @martint and will include your comments as well.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanOutput.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    private void printPlanNodesStatsAndCost(PlanNodeOutput output, PlanNode node)
+    {
+        if (isPlanNodeStatsAndCostsUnknown(node)) {
+            return;
+        }
+
+        output.appendStatsLine(format("Cost: %s", formatPlanNodeStatsAndCost(node)));
+    }
+
+    private boolean isPlanNodeStatsAndCostsUnknown(PlanNode node)
```
[the remark this answers is not in the document]
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanOutput.java]: This is old code that's been moved, but I'll make the change.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeOutput.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class PlanNodeOutput
+{
+    private final String name;
+    private final String identifier;
+    private final String outputs;
+    private final List<PlanNodeId> children;
+    private final StringBuilder details = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeOutput.java]: I can do this if you feel strongly about it, but the concept of 'lines' already seems tied to a textual representation of detail (regardless of whether it's pure text or text as part of GraphViz). Regardless, the API would be 'appendLine' and 'append', i.e. making display-ish decisions.

**Comment by martint** [comment]
I was thinking something along the lines of:

```java
class NodeRepresentation
{
    PlanNodeId id;
    List<PlanNodeId> children;
    String type;
    List<String> outputs;
    Stats stats;
    ...
}

class PlanRepresentation
{
    PlanNodeId root;
    List<NodeRepresentation> node;
}

interface Renderer
{
    render(PlanRepresentation plan);
    render(NodeRepresentation node);
}

class TextRenderer implements Renderer
{
    ...
}


class PlanPrinter
{
    textLogicalPlan(PlanNode root)
    {
        // visit plan
        //    translate PlanNode -> NodeRepresentation
        //    add NodeRepresentation to PlanRepresentation
        // call text renderer
    }
}
```

`TextRenderer.render(NodeRepresentation)` could be used directly to render a node to embed somewhere else (like a verbatim textual representation in the UI). `TextRenderer.render(PlanRepresentation)` knows how to deal with indentation when rendering the tree and delegates to `render(NodeRepresentation)` to render each node.

**Comment by raghavsethi** [comment]
@martint: Updated, please take another look.

Thanks for your suggestions. I chose to not have a way to render a NodeRepresentation, for the following reasons:
1. There is currently no use case for rendering individual nodes - all the use cases render the whole plan.
2. I would have to indent the multiline string returned and use a new StringBuilder for every node. This is not in the fast path so it's not particularly performance sensitive, but given point 1, it doesn't seem to be worth the generated garbage.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanRepresentation.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public int getLevel()
+    {
+        return level;
+    }
+
+    public NodeRepresentation getRoot()
+    {
+        return nodeInfo.get(root.getId());
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanRepresentation.java]: Could you assert that root ID is present in the map to avoid accidentally returning a `null`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanRepresentation.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public Optional<NodeRepresentation> getNode(PlanNodeId id)
+    {
+        return Optional.ofNullable(nodeInfo.get(id));
+    }
+
+    public void addNode(NodeRepresentation nodeOutput)
+    {
+        nodeInfo.put(nodeOutput.getId(), nodeOutput);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanRepresentation.java]: Could you assert that a node with same ID didn't exist before this call to avoid accidentally overwriting it?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        implements Renderer<String>
+{
+    @Override
+    public String render(PlanRepresentation plan)
+    {
+        StringBuilder output = new StringBuilder();
+        return writeTextOutput(output, plan, plan.getLevel(), plan.getRoot());
+    }
+
+    private String writeTextOutput(StringBuilder output, PlanRepresentation plan, int level, NodeRepresentation node)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java]: Can this be a static method?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                .collect(toList());
+
+        for (NodeRepresentation child : children) {
+            writeTextOutput(output, plan, level + 1, child);
+        }
+
+        return output.toString();
+    }
+
+    static String indentString(int indent)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java]: It seems strange that this method is not private.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 10]
```diff
@@ -298,322 +314,106 @@ public static String graphvizDistributedPlan(SubPlan plan)
         return GraphvizPrinter.printDistributed(plan);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: Do you plan to add GraphvizRenderer in a future PR?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java (hunk 9 above)
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java]: It's used in other classes.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 10 above)
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: I don't have that PR planned, but yes that would be the way to go.

**Comment by raghavsethi** [comment]
@mbasmanova Yes you're correct. There are only very minimal changes to the output.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 11]
```diff
@@ -298,322 +314,106 @@ public static String graphvizDistributedPlan(SubPlan plan)
...
-        public Void visitJoin(JoinNode node, Integer indent)
+        public Void visitJoin(JoinNode node, Void context)
         {
             List<Expression> joinExpressions = new ArrayList<>();
             for (JoinNode.EquiJoinClause clause : node.getCriteria()) {
                 joinExpressions.add(clause.toExpression());
             }
             node.getFilter().ifPresent(joinExpressions::add);
 
+            NodeRepresentation nodeOutput;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: s/nodeOutput/representation/

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java (hunk 8 above)
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java]: No, `verbose` is an instance variable.

**Comment by martint** [comment]
Can you post some screenshots of the UI changes?

**Comment by raghavsethi** [comment]
Screenshot

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragment.java [hunk 12]
```diff
@@ -144,6 +147,14 @@ public StatsAndCosts getStatsAndCosts()
         return statsAndCosts;
     }
 
+    @JsonProperty
+    public Optional<String> getJsonRepresentation()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragment.java]: What's the purpose of this field? It's not clear from this commit. Also, it seems odd for this class, which is meant to be a structural representation of the plan, to contain anything related to a rendering/representation for display purposes.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java [hunk 13]
```diff
@@ -36,24 +36,26 @@
         implements Renderer<String>
 {
     private final boolean verbose;
+    private final int level;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/TextRenderer.java]: Is this related to this commit?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanRepresentation.java [hunk 14]
```diff
@@ -27,17 +27,15 @@
 class PlanRepresentation
 {
     private final PlanNode root;
-    private final int level;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanRepresentation.java]: Is this related to this commit?

File: presto-main/src/main/resources/webapp/src/components/QueryDetail.jsx [hunk 15]
```diff
@@ -241,7 +241,7 @@ const HISTOGRAM_PROPERTIES = {
     disableHiddenCheck: true,
 };
 
-class StageDetail extends React.Component {
+class StageSummary extends React.Component {
```
Comment (martint) [inline presto-main/src/main/resources/webapp/src/components/QueryDetail.jsx]: Is this a preexisting issue? If so, can you add details in the commit explaining the impact of this fix? (i.e., was part of the UI broken?). If not, it should be squashed with the relevant commit.

File: presto-main/src/main/resources/webapp/src/components/LivePlan.jsx [hunk 16]
```diff
@@ -247,12 +247,28 @@ export class LivePlan extends React.Component<LivePlanProps, LivePlanState> {
...
+        const inner = d3.select("#plan-canvas g");
+        this.state.render(inner, graph);
 
         const svg = this.state.svg;
         svg.selectAll("g.cluster").on("click", LivePlan.handleStageClick);
-        svg.attr("height", graph.graph().height);
-        svg.attr("width", graph.graph().width);
+
+
+        const graphWidth = graph.graph().width + 100;
```
Comment (martint) [inline presto-main/src/main/resources/webapp/src/components/LivePlan.jsx]: Why "+ 100" ?

File: presto-main/src/main/resources/webapp/src/components/LivePlan.jsx [hunk 17]
```diff
@@ -247,12 +247,28 @@ export class LivePlan extends React.Component<LivePlanProps, LivePlanState> {
...
+        const graphHeight = graph.graph().height + 100;
+        const width = parseInt(window.getComputedStyle(document.getElementById("live-plan"), null).getPropertyValue("width").replace(/px/, "")) - 50;
```
Comment (martint) [inline presto-main/src/main/resources/webapp/src/components/LivePlan.jsx]: These are a little hard to read. I'd, at a minimum, pull out `window.getComputedStyle(document.getElementById("live-plan"), null)` into a variable since it's used in both lines.

Comment (martint) [reply presto-main/src/main/resources/webapp/src/components/LivePlan.jsx]: Also is the `null` in `window.getComputedStyle(document.getElementById("live-plan"), null)` required? Online docs seem to indicate that omitting is is equivalent to passing `null`.
